cloudify/utils.py

```diff
@@ -1058,3 +1063,53 @@ def uuid4():
         uuid_as_hex[16:20],
         uuid_as_hex[20:]
     )
+
+
+def keep_trying_http(total_timeout_sec=KEEP_TRYING_HTTP_TOTAL_TIMEOUT_SEC,
+                     max_delay_sec=MAX_WAIT_BETWEEN_HTTP_RETRIES_SEC):
+    """
+    Keep (re)trying HTTP requests.
+
+    This is a wrapper function that will keep (re)running whatever function it
+    wraps until it raises a "non-repairable" exception (where "repairable"
+    exceptions are `requests.exceptions.ConnectionError` and
+    `requests.exceptions.Timeout`), or `total_timeout_sec` is exceeded.
+
+    :param total_timeout_sec: The amount of seconds to keep trying, if `None`,
+                              there is no timeout and the wrapped function
+                              will be re-tried indefinitely.
+    :param max_delay_sec: The maximum delay (in seconds) between consecutive
+                          calls to the wrapped function.  The actual delay will
+                          be a pseudo-random number between
+                          `0` and `max_delay_sec`.
+    """
+    logger = setup_logger('http_retrying')
+
+    def ex_to_str(exception):
+        return str(exception) or str(type(exception))
+
+    def inner(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            timeout_at = None if total_timeout_sec is None \
+                else datetime.utcnow() + timedelta(seconds=total_timeout_sec)
+            while True:
+                try:
+                    return func(*args, **kwargs)
+                except (ConnectionError, Timeout) as ex:
+                    if timeout_at and datetime.utcnow() > timeout_at:
+                        logger.exception(f'Finished retrying {func}: total '
+                                         f'timeout of {total_timeout_sec} '
+                                         f'seconds exceeded: {ex_to_str(ex)}')
+                        raise
+                    delay = random.randint(0, max_delay_sec)
+                    logger.warning(f'Will retry {func} in {delay} seconds: ' +
+                                   ex_to_str(ex))
+                    time.sleep(delay)
+                except Exception as ex:
+                    logger.exception(f'Will not retry {func}: the encountered '
+                                     'error cannot be fixed by retrying: ' +
+                                     ex_to_str(ex))
```

Comment:
With logging.exception, you don't need to provide the exception to it, you just need to use it in an exception handler and it'll put the traceback at the end after a newline:
`>>> def testit():
...     try:
...         raise ValueError('test')
...     except Exception:
...         logger.exception('There was a bad.')
... 
>>> testit()
There was a bad.
Traceback (most recent call last):
  File "<stdin>", line 3, in testit
ValueError: test
`

Reply:
After some external discussion: These should probably be logger.error anyway, the full traceback is almost certainly just noise given the error types we have.